torch/distributed/_shard/checkpoint/storage.py

```diff
@@ -0,0 +1,167 @@
+import abc
+from typing import List, Dict
+
+from torch.futures import Future
+
+from .metadata import (
+    BytesReadRequest,
+    BytesWriteRequest,
+    Metadata,
+    TensorReadRequest,
+    TensorWriteRequest,
+)
+
+class StorageWriter(abc.ABC):
+    """
+    Interface used by ``save_state_dict`` to write to storage.
+
+    A subclass should expect the following sequence of calls by ``save_state_dict``
+
+    1) prepare_storage with a dict of storage_key -> size. This can be used to
+        pre-allocate storage.
+    2) write_bytes
+    3) write_tensors.
+    4) (if rank == 0) write_metadata().
+
+    Rank 0 will wait for (2) and (3) before writing metadata.
+    The writes from (2) and (3) are initiated before any waiting is done.
```

Comment:
Is 3 called after future in 2 completes? Seems a bit unclear.
What if 2 fails? What's the failure handling story?

Reply:
Updated the doc. The idea is that all writes are waited together and failure of either will invalidate the checkpoint.

The error handling story is going to be properly flushed out after this PR as it needs to be done across all ranks to ensure the checkpoint is correct.